### ISSUE TITLE
memory leak through plumpy's `PROCESS_STACK` context variable

### ISSUE THREAD

**Issue by ltalirz** [aiidateam/aiida-core#4698]
### Describe the bug

plumpy uses the `contextvars` module to enable access to the process itself (via `Process.current()`)  at any point in time during a task run by the process.

https://github.com/aiidateam/plumpy/blob/33f4e9f029e0573f168237b8db030de537453878/plumpy/processes.py#L516-L534

When used in combination with `asyncio`'s `call_soon`, `call_later` or `call_at` methods, each individual function execution actually [gets their own copy of this context](https://www.python.org/dev/peps/pep-0567/#asyncio).
This means that as long as a handle to these scheduled executions remains in memory, the copy of the  `'process stack'` context var (and thus the process itself) remain in memory [1].

One example of this happening is the `Transport` future opened by aiida-core for file transfers - it is passed around to several classes inside aiida-core and even outside (e.g. `paramiko`).
https://github.com/aiidateam/aiida-core/blob/97cecd2ef57946dd53a0ecd2005f3d2d0a94a2aa/aiida/engine/transports.py#L82-L99

The problem is: if any of these places forgets to shed the reference to this handle, not only does the transport itself remain in memory but also the Process that was responsible for creating it (and everything it references).

One can work around this issue by explicitly passing an empty context to `call_soon`, `call_later` or `call_at` where the context is not needed, but I would like to point out that this behaviour may be unexpected to users of plumpy and can be very difficult to debug.

### Your environment

- aiida-core Current `develop`

@muhrin 

[1] Not sure whether memory consumption during execution should be a concern as well - if it's a shallow copy, it should be ok.

**Comment by greschd** [comment]
Maybe the context (or the process stack) should only hold a weak reference to the process? Or is there a use case where the process needs to be kept "in memory" by the context?

It doesn't completely eliminate the possibility of someone keeping the process alive - the weak reference can be evaluated into a "regular" reference. But probably it's less likely to happen in parts of the code that don't need the process in the first place.

```python
In [1]: import weakref

In [2]: class TestObject:
   ...:     """needed because built-in types do not support weak references"""
   ...:

In [3]: f = TestObject()

In [4]: g = weakref.ref(f)

In [5]: g
Out[5]: <weakref at 0x7f219514d048; to 'TestObject' at 0x7f21928fd048>

In [6]: del f

In [7]: g  # dead after f has been deleted
Out[7]: <weakref at 0x7f219514d048; dead>

In [8]: f = TestObject()

In [9]: g = weakref.ref(f)

In [10]: h = g() # turn back into a normal reference

In [11]: del f

In [12]: g
Out[12]: <weakref at 0x7f219333b7c8; to 'TestObject' at 0x7f21928f0ef0>

In [13]: del h

In [14]: g # dead only after both f and h have been deleted
Out[14]: <weakref at 0x7f219333b7c8; dead>
```

**Comment by ltalirz** [comment]
@greschd Thanks, I wasn't aware of the `weakref` option. That sounds like a great solution here.
I'll prepare a PR to plumpy

**Comment by chrisjsewell** [comment]
Err, I'm tentative of using weak references, which could non-deterministically dissappear when they are actually required, leading to potentially just as bad side-effects.
I think it is better to first actually clarify where `Process.current()` and hence these context vars are required, and see if perhaps there is a more "selective" way we can pass them around.

Really the only use for weak references is for caches, that are only kept for performance needs, but won't break the code if they are garbage collected

**Comment by greschd** [comment]
> which could non-deterministically dissappear

Right, if they need to be kept in memory by the context in general I agree it's not a useful solution. I wasn't sure if there might be a different part of the system that has "ownership" of the process; from the description above it seemed like the context _shouldn't_ own the process.

Use cases for weakref are pretty rare, but I wouldn't go so far to say it's _only_ for caches; it can also be useful to avoid circular referencing.

**Comment by chrisjsewell** [comment]
> it can also be useful to avoid circular referencing.

Python already has a cycle-detecting garbage collector though

**Comment by ltalirz** [comment]
> which could non-deterministically dissappear when they are actually required, leading to potentially just as bad side-effects.

Let's put it this way: a broken reference in `Process.current()` will be way easier to debug than a hidden chunk of memory references dragged around by the process context.

If one can make the passing around of the context more explicit/obvious, that would be fine as well.

The only thing I would like to avoid is to have to rely on the diligence of users of plumpy to always be aware of what is going on in the background - in order to avoid surprises like the one we had now in aiida.

**Comment by greschd** [comment]
> should not need it for that

It can be needed if you want to implement a non-trivial `__del__` that needs to be called in the right order.

Anyway, I agree (sort of) with @chrisjsewell, in that we don't know exactly what the consequence of making it a weakref is. We should definitely be careful with that, as it could seriously break things.
Since the automatic memory management hasn't really worked out here, we need to figure out who _should_ own the process, and how best to fix that.

I also didn't understand 100% why the futures stay in memory in the first place, after they have been cancelled.

**Comment by chrisjsewell** [comment]
> in that we don't know exactly what the consequence of making it a weakref is. We should definitely be careful with that, as it could seriously break things.

yep thats my key point; lets not jump out of the frying pan, and into the fire 😉

**Comment by ltalirz** [comment]
fixed for the time being in #4699 
to be fixed more generally as described in https://github.com/aiidateam/aiida-core/discussions/4704#discussion-2575373
